ceph: make root= CRUSH label value configurable

horazont opened the pull request:
**Description of your changes:**

- Add crushRoot field to the ClusterSpec
- Use the CRUSHRoot value as label value for the ``root=`` crush label

**Which issue is resolved by this Pull Request:**

Resolves #4993

**Checklist:**

- [ ] **CommitLint Bot**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [x] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test ceph]
[test full]

horazont commented:
I’m a bit lost in the tests. I’d like to add integration tests, but I’m not sure what the proper way to do that is. A bit of guidance here would be appreciated.

I thought about copying e.g. ``ceph_base_deploy_test.go`` and try to put the CRUSHRoot field into the spec created there. Would that be a viable way?

horazont commented:
@travisn Thank you very much for the feedback. I gave a shot at implementing it.

horazont commented:
I noticed that the CRDs in the ``cluster/examples`` subdirectory are not automatically updated with the new crushRoot property of the cluster. What is the proper way to update those? I kind of assumed that ``make codegen`` would keep them in sync, but that doesn’t seem to be the case. Is there any other automated way?

horazont commented:
When testing it in a test cluster, I found that the crush location of the OSD is indeed set correctly via arguments:

```
    - --crush-location=root=foobar host=managed-k8s-worker-0 region=default zone=nova
```

However, the OSD still crashes:

```
2020-05-15 09:40:24.531 7fd8118d2a80 -1 Falling back to public interface
2020-05-15 09:40:25.263 7fd8118d2a80 -1 osd.0 0 log_to_monitors {default=true}
2020-05-15 09:40:25.271 7fd8118d2a80 -1 osd.0 0 mon_cmd_maybe_osd_create fail: '(22) Invalid argument': (22) Invalid argument
2020-05-15 09:40:25.271 7fd8118d2a80 -1 osd.0 0 init unable to update_crush_location: (22) Invalid argument
```

Deploying the toolbox and looking at the osd tree, something seems to have created the ``root`` label before the OSD has started and that seems to cause issues:

```
[root@rook-ceph-tools-7f96779fb9-xswjh /]# ceph osd tree
ID CLASS WEIGHT TYPE NAME    STATUS REWEIGHT PRI-AFF
-1            0 root default
 0   hdd      0 osd.0          down        0 1.00000
 1   hdd      0 osd.1          down        0 1.00000
 2   hdd      0 osd.2          down        0 1.00000
```

May be a red hering, investigating.

horazont commented:
Okay, so, the ``root default`` which was created by default is indeed the issue.

AFAICT, it is created through this call chain in ceph itself:

1. https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L643
2. https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.h#L1460
3. https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.cc#L4138
4. https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.cc#L4241-L4244

Note that after (4), there’s also https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.cc#L4257 and https://github.com/ceph/ceph/blob/master/src/osd/OSDMap.cc#L4340 which creates the `replicated_rule` crush rule using the root node, so the root node cannot simply be deleted; the `replicated_rule` needs to be deleted first.

I’m not sure what can be done about this. My initial thought is to delete `root=default` and `replicated_rule` when initialising the cluster (maybe after creating the first mon?) and ignore errors when they cannot be deleted (e.g. because they’re in use). If a pool uses the rule or the some other rule uses the root=default node or an OSD or other node is put below the root=default node, deletion will fail.

I’m going to give this a shot.

horazont commented:
Alright, I implemented code which will remove the default ``replicated_rule`` as well as the ``default`` crush node if they exist and are not used. If the node is not empty or if either of the two is in use, they will be left in place.

I made a local test which confirms that the cluster spawns into an operational state.

The integration tests couldn’t have caught this issue because it only becomes an issue if something else attempts to use the ``default`` CRUSH value :/.

jdurgin commented:
> > I noticed that the CRDs in the cluster/examples subdirectory are not automatically updated with the new crushRoot property of the cluster. What is the proper way to update those? I kind of assumed that make codegen would keep them in sync, but that doesn’t seem to be the case. Is there any other automated way?
> 
> @horazont The examples are not updated automatically so you would need to modify the cluster.yaml.
> 
> > Okay, so, the root=default which was created by default is indeed the issue.
> 
> @jdurgin There is a desire to have a root name other than "default". Is there a way to skip creating the "default" CRUSH root or set an alternate root name in instead of deleting it after the cluster is initialized?

You can provide an osdmap when creating the monitor. Namely during --mkfs, you can pass an osdmap file you've generated via osdmaptool as the initial one for the monitor via 'ceph-mon --mkfs --osdmap /path/to/map ...' 

I'm not sure anything uses this option, so it's not regularly tested. It only affects the initial creation of the cluster, for later monitor additions it would have no impact.

horazont commented:
First off, I’d like to mention that I’m going to be on vacation next week and since this is work stuff, I’ll probably not be able to do any code fixes during that time. 

> > I noticed that the CRDs in the cluster/examples subdirectory are not automatically updated with the new crushRoot property of the cluster. What is the proper way to update those? I kind of assumed that make codegen would keep them in sync, but that doesn’t seem to be the case. Is there any other automated way?
> 
> @horazont The examples are not updated automatically so you would need to modify the cluster.yaml.

Ohkay. I was referring to common.yaml with all the CustomResourceDefinition resources though. Is that maintained manually or is there a way to generate that from the struct types?

horazont commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -391,5 +437,11 @@ func (c *cluster) postMonStartupActions() error {
 
+	// Remove the root=default and replicated_rule which are created by default. Note that RemoveDefaultCrushMap
+	// ignores some types of errors internally
+	if err := c.removeDefaultCrushMap(); err != nil {
```
Hm, when I ran my tests locally, I found that the ``ceph osd crush rm`` command fails instead of breaking anything if the node is still in use (e.g. because things are nested under it).

Based on that assumption, I think that it’s not a problem to remove that node and the rule always if they’re unused. Was that unclear or do you still think that the removal should be guarded for first_run && crushRoot != default?

horazont replied:
@travisn Hm, I tried to mention in the comment that the removeDefaultCrushMap function will ignore certain errors (in this case, EBUSY and ENOTEMPTY) internally already. The remaining errors are more fatal (e.g. if the ceph CLI can’t reach the mons or something) and those should IMO fail loudly.

If you disagree on that, I’m going to change it though.

horazont commented on cluster/examples/kubernetes/ceph/cluster.yaml:
```diff
@@ -184,3 +184,6 @@ spec:
     machineDisruptionBudgetNamespace: openshift-machine-api
+  # If any of your topology labels or other input to the CRUSH map labels have the value `default`, you need to change this to avoid
+  # conflicts.
+  crushRoot: default
```
Could do, if you think that it makes more sense. I’m normally in favour of putting things in schema.

horazont commented:
@travisn 

- Rebased on master
- Moved crushRoot from the Spec root to the storage.config
- Changed the code so that it’ll only touch root=default if and only if the crushRoot was set.

leseb commented on cmd/rook/ceph/osd.go:
```diff
@@ -241,8 +241,13 @@ func commonOSDInit(cmd *cobra.Command) {
 func getLocation(clientset kubernetes.Interface) (string, error) {
 	// get the value the operator instructed to use as the host name in the CRUSH map
 	hostNameLabel := os.Getenv("ROOK_CRUSHMAP_HOSTNAME")
+	rootLabel := os.Getenv("ROOK_CRUSHMAP_ROOT")
```
Can't we always populate this with a value instead? So if `osdProps.crushRoot` is empty then we set `default`. I see multiple places use the same check and I believe we can solve this in a single place, when we set the `ROOK_CRUSHMAP_ROOT` basically.

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,49 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+func (c *cluster) removeDefaultCrushMap() error {
+	logger.Info("removing default objects, if applicable")
+	args := []string{"osd", "crush", "rule", "rm", "replicated_rule"}
+	cephCmd := client.NewCephCommand(c.context, c.Namespace, args)
```
Use the context from `c` and the executor to run a command instead. I know it might be confusing because the code base is not homogenous yet...

horazont replied:
@leseb Thanks. I tried that, and in contrast to the client.NewCephCommand way, c.context.Executor will not auto-add the arguments required for the command to succeed, such as keyring and config paths.

I can of course do all that manually, but duplicating the work of https://github.com/rook/rook/blob/master/pkg/daemon/ceph/client/command.go#L56-L88 seems not ideal. Calling that function would of course also be an option.

Which way is the preferred one?

1. c.context.Executor + FinalizeCephCommandArgs + whatever else may be needed
2. c.context.Executor + pass only the strictly necessary args by composing them manually
3. stay with client.NewCephCommand

leseb replied:
Actually my bad, `NewCephCommand` should be used here since it's building the correct arguments. So please dismiss my earlier comment. The executor should be used for non-ceph commands only :).

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,49 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+		}
+	}
+
+	args = []string{"osd", "crush", "rm", "default"}
```
Please create a separate method.

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,49 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+		if exitErr, ok := err.(*exec.ExitError); ok {
+			if exitErr.ExitCode() != int(syscall.EBUSY) {
+				// the error does not refer to EBUSY -> return as error
+				return err
```
Name your errors with  `errors.Wrap`.

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,49 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+				// created by the user with a different crush location.
+			}
+		} else {
+			return err
```
ditto

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,49 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+			if exitErr.ExitCode() != int(syscall.ENOTEMPTY) && exitErr.ExitCode() != int(syscall.EBUSY) {
+				// the error does not refer to EBUSY or ENOTEMPTY
+				// -> return as error
+				return err
```
ditto

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,49 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+				logger.Debug("default is not empty or is still in use, not removed")
+			}
+		} else {
+			return err
```
ditto

leseb commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -775,10 +786,28 @@ func getLocationFromPod(clientset kubernetes.Interface, d *apps.Deployment) (str
+		}
+
+		for _, envvar := range container.Env {
+			if envvar.Name == "ROOK_CRUSHMAP_ROOT" {
```
Please create a constant for `ROOK_CRUSHMAP_ROOT` since it's called multiple times.

leseb commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -775,10 +786,28 @@ func getLocationFromPod(clientset kubernetes.Interface, d *apps.Deployment) (str
 }
 
-func GetLocationWithNode(clientset kubernetes.Interface, nodeName string, crushHostname string) (string, error) {
+func GetLocationWithNode(clientset kubernetes.Interface, nodeName string, crushRoot string, crushHostname string) (string, error) {
```
```suggestion
func GetLocationWithNode(clientset kubernetes.Interface, nodeName, crushRoot, crushHostname string) (string, error) {
```

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,66 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
 }
 
+func (c *cluster) removeDefaultReplicationRule() error {
+	err := c.context.Executor.ExecuteCommand("ceph", "osd", "crush", "rule", "rm", "replicated_rule")
```
Sorry I led you the wrong way, we need to revert this, `git ref` to the rescue! You can still use the context though but not its `Executor`.

horazont commented:
@travisn @leseb Thanks again for the useful feedback.

- Cleaned up the default CRUSH map removal functions (reverting back to NewCephCommand, too)
- Removed the explicit env var lookup in cluster/osd/osd.go
- Added more explicit comments in the removal functions explaining why EBUSY/ENOTEMPTY are ignored

leseb commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -67,6 +70,16 @@ type cluster struct {
+	if v, ok := c.Storage.Config[crushRootConfigKey]; ok {
+		return v
+	}
+	return ""
```
Why not returning `default` here? So that we will always populate it and we don't need to check if empty in other portion of the code?

leseb commented on pkg/operator/ceph/cluster/osd/osd_test.go:
```diff
@@ -392,3 +392,46 @@ func TestGetOSDInfo(t *testing.T) {
+	assert.Equal(t, osd2.ID, osds2[0].ID)
+	assert.Equal(t, osd2.BlockPath, osds2[0].BlockPath)
+	assert.Equal(t, osd2.CVMode, osds2[0].CVMode)
+	assert.Equal(t, osd2.LVBackedPV, osds2[0].LVBackedPV)
```
shouldn't you test `location` contains `root-default` here?

horazont commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,86 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+
+	err = c.removeDefaultReplicationRule()
+	if err != nil {
+		return err
```
@leseb Hm, is that a policy?

The errors returned by the methods are already wrapped, and I can’t think of any string for use with errors.Wrap which isn’t redundant:

- "failed to remove default CRUSH map" is redundant with the errors.Wrap which is already happening at the call site in postMonStartupActions
- "failed to remove default replication_rule rule" is redundant with the wrapping in removeDefaultReplicationRule

Maybe I’m using this wrong?

leseb replied:
I know that in some case it's redundant but this is a practice we are trying to enforce since it makes debugging simpler. 

What about `failed to remove default rule`? and `removeDefaultReplicationRule()` is more precise since it names the rule.

horazont commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,86 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+			//   CRUSH root
+			// in both cases, we cannot do anything about it either way and
+			// we’ll assume that the user knows what they’re doing.
+			logger.Debug("replicated_rule is in use, not removed")
```
@leseb Ok, but I’d then suggest at most Warning level, because the rule `replicated_rule` could’ve been (re-)created by the user using a different CRUSH root, so it’s not actually a problem. Logging with Error could be misleading and confusing.

leseb replied:
Ok Warning sounds good then.

horazont commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -397,6 +411,86 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
+				//
+				// in all cases, we cannot do anything about it either way and
+				// we’ll assume that the user knows what they’re doing.
+				logger.Debug("default is not empty or is still in use, not removed")
```
@leseb Same as above. This message will *always* appear in a cluster which uses a non-default root for a good reason, since the `osd crush rm` command only goes by "value" and not by "level": We can’t say `osd crush rm root=default`, but can only say `osd crush rm default`.

That means that when the non-default root is chosen because, e.g. the region topology key is `default`, every cluster operation will cause this message to be logged; the operator will execute this and fail to `osd crush rm default`, because the (e.g.) `region=default` CRUSH node is not empty or still in use.

An error would be even more confusing, IMO, than in the other case.

Sorry if this was obvious. I just want to make sure we’re on the same page.

horazont commented:
Sorry for the delay, this got lost in the holiday weekend. I’ll get to this tomorrow.

horazont commented:
Pushed another round of fixes.

@leseb @travisn I took the liberty to resolve all conversations which I think I addressed. The list got quite confusing to me to keep track of. I hope that’s alright.

@leseb There are a few points which I didn’t address yet since I’m waiting for feedback.

horazont commented:
> @horazont Did you see my comment earlier about the integration tests failing? Looks like the default case is not working.

@travisn Yes, I rebased my branch and tried a fix and waited for the CI result. I still need to understand what exactly is going wrong though (I don’t quite understand the structure of the CI output). I may have to spin up a deployment locally tomorrow to see where it breaks.

horazont commented:
@travisn Sorry for the delays… Stuff is getting in the way.

So. I tried another fix, but it still doesn’t work (obviously). The Jenkins UI is highly non-obvious for me, can you give me a pointer where to look for logs from the different rook components? (Are they even included?)

galexrt commented:
@horazont The logs of each Pod and the nodes can be found in the "Artifacts" tab at the top of the CI page, see https://jenkins.rook.io/blue/organizations/jenkins/rook%2Frook/detail/PR-5483/16/artifacts

There you can download the logs of each Pod and even dmesg, and other logs if needed.

horazont commented:
Not yet, going to do that soon.

leseb commented:
@horazont any plan to pursue this work? Thanks

horazont commented:
@leseb  "Yes, but". I’ve been pulled away from this at work, I’m trying to get some time scheduled to finish this.

horazont commented:
@leseb @travisn I finally got time alotted and fixed the issue. I spawned two lab clusters, one with a default crush root (unset via CRD) and one with a non-default crush root.

Both seem to work fine. I had to change a few things to catch up with current master, so I suppose you both might want to re-review, hence the explicit ping.

horazont commented:
A hint on what may be wrong in the CI would be appreciated :/

horazont commented:
I found a problem with the devicehealth module, I’ll have to investigate how to fix it.

The devicehealth wants to create a pool, but it doesn’t specify a crush rule. Since we delete the default rules, the module cannot succeed.

leseb commented:
> I found a problem with the devicehealth module, I’ll have to investigate how to fix it.
> 
> The devicehealth wants to create a pool, but it doesn’t specify a crush rule. Since we delete the default rules, the module cannot succeed.

Can't we change the rules after it's created and do our rules operations after?

horazont commented:
So instead of deleting the default rule, change it to point to root=whateverwasconfigured? That sounds like a decent plan, I’ll look into that.

horazont commented:
So this was another fun yak shaving party. While fixing the CRUSH rules, I realized that I hadn’t taken care of block pools, filesystems and object storage yet. Those also want to use the correct root by default, instead of a random different non-root root (which would silently work in most cases, since you presumably have a good reason to choose a non-default root crush label).

Eiiitherway. I took care of that now. Again a larger change compared to the previous diff. In addition, I haven’t squashed the commits yet for easier reviewing of that part of the change. I’ll do that once I get greenlight, since I expect to have to rebase before this can hit master anyways.

I did the following tests in a test cluster where the OpenStack region=default:

- [x] cluster spawn on PVC with root=non-default
- [x] create block pool. write data in one pod, create another pod (on a different worker, thus forcing remount), read and verify data
- [x] create shared filesystem. create three reader/writer pods which create files and wait for files from the other pods to appear.
- [x] verified that all created pools have the correct crush root

I’ll do the same tests on a cluster where root=default and the OpenStack region is not default.

horazont commented on pkg/operator/ceph/file/controller.go:
```diff
@@ -222,9 +222,11 @@ func (r *ReconcileCephFilesystem) reconcile(request reconcile.Request) (reconcil
 		return reconcile.Result{}, errors.Wrapf(err, "invalid object filesystem %q arguments", cephFilesystem.Name)
 	}
 
+	crushRoot := opconfig.GetCrushRootFromSpec(&cephCluster.Spec)
```
I thought about that too, but decided against it. I think overriding the CRUSH root for a pool is a different use-case than overriding the CRUSH root for the cluster.

Overriding the CRUSH root for a pool may put a pool in a specific availability zone for example. Changing the CRUSH root for the cluster should be mostly invisible and only be used to resolve conflicts between auto-created CRUSH labels and the root=default label.

So my argument would be that for the sake of user experience to keep this special handling and treat absent crushRoot in a pool to use the crush root form the cluster, because they serve different use-cases. When not setting a CRUSH root on a pool, it is expected that it uses the same as the cluster, IMO.

If we are to change this behaviour, I would like to be able to error-out if the crush root is not specified on the pool && the cluster uses a non-default crush root (otherwise, weird things will happen I’m afraid). This would again require looking up the crush root from the cluster, so there’s not much simplification either way.

horazont replied:
@travisn I went with the better UX. I don’t think we’d gain much by changing this in the code. Feel free to overrule (in which case I’ll change it).

horazont commented:
Thanks for the additional feedback. I’m on vacation currently and will hopefully get around to handle this in a week or two.

galexrt commented:
@horazont Ping. The PR has merge conflicts, are you available to take another look at your PR?

horazont commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -246,6 +247,7 @@ func (c *Cluster) startProvisioningOverPVCs(config *provisionConfig) {
 			placement:        volume.Placement,
 			portable:         volume.Portable,
 			crushDeviceClass: volume.CrushDeviceClass,
+			crushRoot:        opconfig.GetCrushRootFromSpec(&c.spec),
```
@travisn Good idea. I did that for the OSD functions.

For the Pool functions (in the client package) I decided against it because they do not yet receive the cluster spec and requiring them to receive that seems odd to me (as they really only need the cluster crush root). If you think differently on this, let me know.

horazont commented:
@travisn Hey, I think I addressed the things you needed. The CI also looks good to me now. Is there anything else I need to do now?

horazont commented on pkg/operator/ceph/file/filesystem.go:
```diff
@@ -174,18 +176,18 @@ func newFS(name, namespace string) *Filesystem {
 	// generating the metadata pool's name
 	metadataPoolName := generateMetaDataPoolName(f)
-	err := client.CreatePoolWithProfile(context, clusterInfo, metadataPoolName, spec.MetadataPool, "")
+	err := client.CreatePoolWithProfile(context, clusterInfo, metadataPoolName, spec.MetadataPool, "", clusterCrushRoot)
```
@travisn As pointer/reference or directly?

horazont replied:
Ok, I went for "pass-by-value". I also had to move the GetCrushRootFromSpec to the client package in order to avoid a cyclic import.

horazont commented on pkg/operator/ceph/cluster/cluster.go:
```diff
@@ -373,6 +375,116 @@ func (c *ClusterController) preClusterStartValidation(cluster *cluster, clusterO
 }
 
+func extractExitCode(err error) (int, bool) {
+	exitErr, ok := err.(*exec.ExitError)
```
Nope, didn’t touch that piece. Must’ve been some update, e.g. through the rebase.

The only traceback related to ExitError I can find in the run you linked does not relate to my code either:

```
--- FAIL: TestCephHelmSuite (19.58s)
    suite.go:62: test panicked: interface conversion: error is *exec.Error, not *exec.ExitError
        goroutine 653 [running]:
        runtime/debug.Stack(0xc0008ab528, 0x184a640, 0xc000826960)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
        github.com/stretchr/testify/suite.failOnPanic(0xc000836800)
        	/home/rook/go/pkg/mod/github.com/stretchr/testify@v1.4.0/suite/suite.go:62 +0x57
        panic(0x184a640, 0xc000826960)
        	/usr/local/go/src/runtime/panic.go:679 +0x1b2
        github.com/rook/rook/pkg/util/exec.runCommandWithOutput(0xc0009278c0, 0x0, 0xc0009437c0, 0x2, 0x2, 0xc0009278c0)
        	/home/rook/go/src/github.com/rook/rook/pkg/util/exec/exec.go:285 +0x2b2
        github.com/rook/rook/pkg/util/exec.(*CommandExecutor).ExecuteCommandWithOutput(0x2b38218, 0x1a5864b, 0x4, 0xc0009437c0, 0x2, 0x2, 0x10, 0x0, 0x18, 0x20)
        	/home/rook/go/src/github.com/rook/rook/pkg/util/exec/exec.go:128 +0xa0
        github.com/rook/rook/tests/framework/utils.(*HelmHelper).Execute(0xc00091ede0, 0xc0009437c0, 0x2, 0x2, 0x10, 0x0, 0x0, 0x0)
        	/home/rook/go/src/github.com/rook/rook/tests/framework/utils/helm_helper.go:42 +0x7c
        github.com/rook/rook/tests/framework/utils.(*HelmHelper).GetLocalRookHelmChartVersion(0xc00091ede0, 0x1a66bfc, 0xf, 0x1, 0x8, 0xc000583878, 0x10fd01e)
        	/home/rook/go/src/github.com/rook/rook/tests/framework/utils/helm_helper.go:55 +0xa6
        github.com/rook/rook/tests/framework/installer.(*CephInstaller).CreateRookOperatorViaHelm(0xc00074f0e0, 0x1a5bd64, 0x7, 0x1a7a146, 0x1b, 0x1, 0x1)
        	/home/rook/go/src/github.com/rook/rook/tests/framework/installer/ceph_installer.go:179 +0x6e
        github.com/rook/rook/tests/framework/installer.(*CephInstaller).InstallRook(0xc00074f0e0, 0x1a5bd64, 0x7, 0x1a5e602, 0x9, 0x2860d00, 0x1a57648, 0x0, 0x1, 0xc000583b01, ...)
        	/home/rook/go/src/github.com/rook/rook/tests/framework/installer/ceph_installer.go:462 +0x129
        github.com/rook/rook/tests/integration.(*TestCluster).Setup(0xc0002dde00)
        	/home/rook/go/src/github.com/rook/rook/tests/integration/ceph_base_deploy_test.go:169 +0xd3
        github.com/rook/rook/tests/integration.StartTestCluster(0xc00084d930, 0xc0002dde00, 0x1a51820, 0x409379)
        	/home/rook/go/src/github.com/rook/rook/tests/integration/ceph_base_deploy_test.go:161 +0x303
        github.com/rook/rook/tests/integration.(*HelmSuite).SetupSuite(0xc0008a00c0)
        	/home/rook/go/src/github.com/rook/rook/tests/integration/ceph_helm_test.go:80 +0x15f
        github.com/stretchr/testify/suite.Run(0xc000836800, 0x1d66500, 0xc0008a00c0)
        	/home/rook/go/pkg/mod/github.com/stretchr/testify@v1.4.0/suite/suite.go:102 +0x4ff
        github.com/rook/rook/tests/integration.TestCephHelmSuite(0xc000836800)
        	/home/rook/go/src/github.com/rook/rook/tests/integration/ceph_helm_test.go:52 +0xae
        testing.tRunner(0xc000836800, 0x1b68a20)
        	/usr/local/go/src/testing/testing.go:909 +0xc9
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:960 +0x350
```

Looking at the docs of [Error](https://golang.org/pkg/os/exec/#Error), it seems to me as something in the environment is borked, maybe the `ceph` executable is missing?

(Given that I use `foo, ok := bar.(*type)` and not `foo := bar.(*type)`, I would also expect `ok` to simply be false instead of a panic.)

horazont replied:
@travisn So this is not related to my changes? Should I just wait for a fix and rebase afterward?

leseb replied:
Fixed here https://github.com/rook/rook/pull/6503

leseb commented:
Hum I think @travisn clicked the wrong button.

leseb commented:
> Hum I think @travisn clicked the wrong button.

Oh no it's just github I guess :-(
